# Pipeline for experiments (VG)

## NOTE: It is assumed that you have `human_v38.fa` reference genome and `hprc-v1.0-pggb.grch38.1-22+X.vcf` HPRC variant calls files. 
## NOTE: You can get HPRC data from:
##          https://s3-us-west-2.amazonaws.com/human-pangenomics/index.html?prefix=pangenomes/freeze/freeze1/pggb/vcfs/
## NOTE: Pacbio-HiFi reads can be found at:
##          https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/HG002/assemblies/polishing/HG002/v1.0/mapping/hifi_revio_pbmay24/hg002v1.0.1_hifi_revio_pbmay24.bam
## NOTE: ONT reads can be found at:
##          https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/HG002/assemblies/polishing/HG002/v1.0/mapping/ont_r10_ul_dorado/hg002v1.0_ont_r10_ul_dorado.bam
## NOTE: VCF generated by PBSV can be found at:
##          https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/data/AshkenazimTrio/analysis/PacBio_pbsv_05212019/HG002_GRCh38.pbsv.vcf.gz
## Processing step for reads:
##  HiFi:
##      samtools view -b hg002v1.0.1_hifi_revio_pbmay24.bam chr1_MATERNAL chr1_PATERNAL chr10_MATERNAL chr10_PATERNAL chr22_MATERNAL chr22_PATERNAL | samtools fastq - > hg002v1.0.1_hifi_revio_pbmay24.chr1_10_22.fastq
##      sed -n '1~4s/^@/>/p;2~4p' hg002v1.0.1_hifi_revio_pbmay24.chr1_10_22.fastq > hg002v1.0.1_hifi_revio_pbmay24.chr1_10_22.fa
##      rm hg002v1.0.1_hifi_revio_pbmay24.chr1_10_22.fastq
##      seqtk sample -s100 hg002v1.0.1_hifi_revio_pbmay24.chr1_10_22.fa 0.45 > hg002v1.0.1_hifi_revio_pbmay24.chr1_10_22.subsampled.fa
##      
##      samtools view -b hg002v1.0.1_hifi_revio_pbmay24.bam | samtools fastq - > hg002v1.0.1_hifi_revio_pbmay24.fastq
##      sed -n '1~4s/^@/>/p;2~4p' hg002v1.0.1_hifi_revio_pbmay24.fastq > hg002v1.0.1_hifi_revio_pbmay24.fa
##      rm hg002v1.0.1_hifi_revio_pbmay24.fastq
##      seqtk sample -s100 hg002v1.0.1_hifi_revio_pbmay24.fa 0.45 > hg002v1.0.1_hifi_revio_pbmay24.subsampled.fa
##      mv hg002v1.0.1_hifi_revio_pbmay24.subsampled.fa hg002v1.0.1_hifi_revio_pbmay24.fa
##  ONT:
##      samtools view -b hg002v1.0_ont_r10_ul_dorado.bam chr1_MATERNAL chr1_PATERNAL chr10_MATERNAL chr10_PATERNAL chr22_MATERNAL chr22_PATERNAL | samtools fastq - > hg002v1.0_ont_r10_ul_dorado.chr1_10_22.fastq
##      sed -n '1~4s/^@/>/p;2~4p' hg002v1.0_ont_r10_ul_dorado.chr1_10_22.fastq > hg002v1.0_ont_r10_ul_dorado.chr1_10_22.fa
##      rm hg002v1.0_ont_r10_ul_dorado.chr1_10_22.fastq
##      seqtk sample -s100 hg002v1.0_ont_r10_ul_dorado.chr1_10_22.fa 0.25 > hg002v1.0_ont_r10_ul_dorado.chr1_10_22.subsampled.fa
##
## The required data:
##      ├── human_v38.fa
##      ├── human_v38.fa.fai
##      ├── hprc-v1.0-pggb.grch38.1-22+X.vcf
##      ├── hprc-v1.0-pggb.grch38.1-22+X.vcf.gz
##      ├── hprc-v1.0-pggb.grch38.1-22+X.vcf.gz.tbi
##      ├── HG002_GRCh38.pbsv.vcf.gz
##      ├── HG002_GRCh38.pbsv.vcf.gz.tbi
##      └── reads/
##          ├── hg002v1.0.1_hifi_revio_pbmay24.fa
##          ├── hg002v1.0.1_hifi_revio_pbmay24.chr1_10_22.subsampled.fa
##          └── hg002v1.0_ont_r10_ul_dorado.chr1_10_22.subsampled.fa
##
## NOTE: samtools is required (system wide installed) for gaf to sam converion

# MODES
: "${LOCAL_TMP:=true}"
: "${LCPAN_RGFA:=true}"
: "${LCPAN_GFA:=true}"
: "${LCPAN_THD:=true}"
: "${VG_THD:=true}"
: "${ALIGN_LIMITED:=true}"
: "${ALIGN:=true}"

# Directories and files
# Program will copy the files to the 
AKHAL_DIR=
GRAPHALIGNER_DIR=
LCPAN_DIR=
LCPANMERGE_DIR=

FASTA=
FAI=
HPRC=
HPRC_GZ=
HPRC_GZ_TBI=

HG002_VCF=
PACBIO_FA=
PACBIO_SUB_FA=
ONT_SUB_FA=

if [ "$LOCAL_TMP" = "true" ]; then
    cd /tmp
fi

mkdir -p lcpan
cd lcpan

### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------
### Prepare DATA
### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------

if [ "$LOCAL_TMP" = "true" ]; then

    FILE_FASTA=$(basename $FASTA)
    FILE_FAI=$(basename $FAI)
    FILE_HPRC=$(basename $HPRC)
    FILE_HPRC_GZ=$(basename $HPRC_GZ)
    FILE_HPRC_GZ_TBI=$(basename $HPRC_GZ_TBI)
    FILE_HG002_VCF=$(basename $HG002_VCF)
    FILE_PACBIO_FA=$(basename $PACBIO_FA)
    FILE_PACBIO_SUB_FA=$(basename $PACBIO_SUB_FA)
    FILE_ONT_SUB_FA=$(basename $ONT_SUB_FA)

    if [ ! -f $FILE_FASTA ]; then
        cp $FASTA .
    fi
    FASTA="/tmp/lcpan/$FILE_FASTA"
    if [ ! -f $FILE_FAI ]; then
        cp $FAI .
    fi
    FAI="/tmp/lcpan/$FILE_FAI"
    if [ ! -f $FILE_HPRC ]; then
        cp $HPRC .
    fi
    HPRC="/tmp/lcpan/$FILE_HPRC"
    if [ ! -f $FILE_HPRC_GZ ]; then
        cp $HPRC_GZ .
    fi
    HPRC_GZ="/tmp/lcpan/$FILE_HPRC_GZ"
    if [ ! -f $FILE_HPRC_GZ_TBI ]; then
        cp $HPRC_GZ_TBI .
    fi
    HPRC_GZ_TBI="/tmp/lcpan/$FILE_HPRC_GZ_TBI"
    if [ ! -f $FILE_HG002_VCF ]; then
        cp $HG002_VCF .
        cp "$HG002_VCF.tbi" .
    fi
    HG002_VCF="/tmp/lcpan/$FILE_HG002_VCF"

    mkdir -p reads

    cd reads

    if [ "$ALIGN_LIMITED" = "true" ]; then
        if [ ! -f $FILE_PACBIO_SUB_FA ]; then
            cp $PACBIO_SUB_FA .
        fi
        PACBIO_SUB_FA="/tmp/lcpan/reads/$FILE_PACBIO_SUB_FA"
        if [ ! -f $FILE_ONT_SUB_FA ]; then
            cp $ONT_SUB_FA .
        fi
        ONT_SUB_FA="/tmp/lcpan/reads/$FILE_ONT_SUB_FA"
    fi
    if [ "$ALIGN" = "true" ]; then
        if [ ! -f $FILE_PACBIO_FA ]; then
            cp $PACBIO_FA .
        fi
        PACBIO_FA="/tmp/lcpan/reads/$FILE_PACBIO_FA"
    fi

    cd ..
fi

### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------
### HELPER FUNCTIONS
### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------

time_to_seconds() {
    IFS=: read -r -a parts <<< "$1"
    if [ ${#parts[@]} -eq 3 ]; then
        h=${parts[0]}; m=${parts[1]}; s=${parts[2]}
    elif [ ${#parts[@]} -eq 2 ]; then
        h=0; m=${parts[0]}; s=${parts[1]}
    else
        echo "0"; return
    fi
    echo "$h*3600 + $m*60 + $s" | bc -l
}

file_size() {
    local file=$1

    local file_b=$(stat --format="%s" $file 2>/dev/null)
    echo "scale=2; $file_b/(1024*1024*1024)" | bc -l
}

parse_file() {
    local log=$1
    local out=$2

    echo "$log stats:" >> $out

    mapfile -t time < <(grep "Elapsed (wall clock) time" "$log" | awk '{print $8}')
    mapfile -t ram  < <(grep "Maximum resident set size (kbytes)" "$log" | awk '{print $6}')

    if (( ${#time[@]} < 1 || ${#ram[@]} < 1 )); then
        echo "Could not parse $log" >> $out; return
    fi

    local exec_time=$(time_to_seconds "${time[0]}")
    exec_time=$(printf "%.0f" "$exec_time")
    local ram_gb=$(echo "scale=2; ${ram[0]}/(1024*1024)" | bc -l)

    printf -v pretty_time '%02d:%02d:%02d' $((exec_time/3600)) $((exec_time%3600/60)) $((exec_time%60))

    echo "Stats for $log" >> $out
    echo "Alignment time (hh:mm:ss): $pretty_time" >> $out
    echo "RAM usage (GB): $ram_gb" >> $out
    echo "" >> $out
}

parse_2files() {
    # it gets first two /bin/time -v results and ignores the rest (construct and merge)
    local file=$1
    local log=$2
    local out=$3

    echo "$file stats:" >> $out

    mapfile -t times < <(grep "Elapsed (wall clock) time" "$log" | awk '{print $8}')
    mapfile -t rams  < <(grep "Maximum resident set size (kbytes)" "$log" | awk '{print $6}')

    if (( ${#times[@]} < 2 || ${#rams[@]} < 2 )); then
        echo "Could not parse $log" >> $out; return
    fi

    local exec_time=$(time_to_seconds "${times[0]}")
    local merge_time=$(time_to_seconds "${times[1]}")
    local total_time=$(echo "$exec_time + $merge_time" | bc -l)
    local max_ram_kb=$(printf "%s\n%s\n" "${rams[0]}" "${rams[1]}" | sort -nr | head -n1)
    local max_ram_gb=$(echo "scale=2; $max_ram_kb/(1024*1024)" | bc -l)
    local file_gb=$(file_size "$file")

    echo "Execution time (s): $exec_time" >> $out
    echo "Merge time (s): $merge_time" >> $out
    echo "Total runtime (s): $total_time" >> $out
    echo "Max RAM usage (GB): $max_ram_gb" >> $out
    echo "File size (GB): $file_gb" >> $out
    echo "" >> $out
}

fa_stats() {
    local fasta=$1
    local total_len=$2
    local out=$3

    echo "$fasta stats:" >> $out

    awk -v total="$total_len" 'NR % 2 == 0 {
        count++
        sum += length($0)
    }
    END {
        if (count > 0) {
            mean = sum / count
            depth = sum / total
            printf "Total reads: %d\nAverage read length: %.2f\nDepth: %.2f\n", count, mean, depth
        } else {
            print "No reads found!"
        }
    }' "$fasta" >> $out
    echo "" >> $out
}

graphaligner_run() {
    local graph=$1
    local fa=$2
    local reads=$3
    local hg38=$4
    local hg002=$5
    local out="align.$hg002.$reads.$graph.out"

    /bin/time -v ${GRAPHALIGNER_DIR}/GraphAligner \
        -g "$hg38.pggb.$graph.gfa" \
        -f "$fa" \
        -a "$hg002.$reads.$graph.gaf" \
        -x vg \
        -t 96 > $out 2>&1
    
    parse_file "$out" "stats.txt"
}

pbsv_run() {
    local graph=$1
    local fa=$2
    local reads=$3  
    local hg38=$4
    local hg002=$5
    local out="pbsv.$hg002.$reads.$graph.out"

    /bin/time -v ${AKHAL_DIR}/akhal gaf2sam \
        "$hg38.pggb.$graph.gfa" \
        "$hg002.$reads.$graph.gaf"  \
        "$fa" \
        "$hg002.$reads.$graph.sam" \
        --simple > "$out" 2>&1
    
    /bin/time -v ${AKHAL_DIR}/akhal sampoke \
        "$hg38.fa" \
        "$hg002.$reads.$graph.sam" \
        "$hg002.$reads.$graph.qual.sam" >> "$out" 2>&1
    rm -f "$hg002.$reads.$graph.sam"

    samtools view -bo "$hg002.$reads.$graph.bam" "$hg002.$reads.$graph.qual.sam"
    rm -f "$hg002.$reads.$graph.qual.sam"

    samtools sort -@ 16 -o "$hg002.$reads.$graph.sorted.bam" "$hg002.$reads.$graph.bam" 
    rm -rf "$hg002.$reads.$graph.bam"
    mv "$hg002.$reads.$graph.sorted.bam" "$hg002.$reads.$graph.bam"
    samtools index -@ 16 "$hg002.$reads.$graph.bam"
    
    # Perform SV calling
    pbsv discover "$hg002.$reads.$graph.bam" "$hg002.$reads.$graph.svsig.gz"
    pbsv call -j 16 -t DEL,INS,INV -m 20 -A 3 -O 3 --call-min-read-perc-one-sample 20 "$hg38.fa" "$hg002.$reads.$graph.svsig.gz" "$hg002.$reads.$graph.vcf"
    rm -f "$hg002.$reads.$graph.svsig.gz"

    bcftools query -f '%CHROM\t%POS0\t%END\n' "$hg002.$reads.$graph.vcf" > "$hg002.$reads.$graph.bed"

    local GOLD="HG002_GRCh38.chr1_10_22.pbsv.expanded.bed"
    local GOLD_COUNT=$(wc -l < "$GOLD")
    local PRED="$hg002.$reads.$graph.bed"
    
    local FN=$(bedtools intersect -a "$GOLD" -b "$PRED" -v | wc -l)
    local TP=$((GOLD_COUNT - FN))
    local FP=$(bedtools intersect -v -a "$PRED" -b "$GOLD" | wc -l)

    local PRECISION=$(echo "scale=5; $TP / ($TP + $FP)" | bc)
    local RECALL=$(echo "scale=5; $TP / ($TP + $FN)" | bc)
    local F1=$(echo "scale=5; 2 * $PRECISION * $RECALL / ($PRECISION + $RECALL)" | bc)

    echo -e "$reads.$graph\t${TP}\t${FP}\t${FN}\t0${PRECISION}\t0${RECALL}\t0${F1}" >> "stats.txt"
}

### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------
### LCP levels analyses based on non-overlapping rgfa graph
### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------

if [ "$LCPAN_RGFA" = "true" ]; then
    echo "Experiment on different LCP levels started (nov-rgfa)..."

    mkdir -p lcpan-levels-nov-rgfa
    cd lcpan-levels-nov-rgfa

    for l in 4 5 6 7; do \
        /bin/time -v ${LCPAN_DIR}/lcpan -vg \
            -r "$FASTA" \
            -v "$HPRC" \
            -p hg38.pggb.lcpan.l${l} \
            -l ${l} \
            --verbose > hg38.pggb.lcpan.l${l}.out 2>&1; \
        /bin/time -v bash ${LCPANMERGE_DIR}/lcpan-merge.sh hg38.pggb.lcpan.l${l}.log >> hg38.pggb.lcpan.l${l}.out 2>&1; \
    done
    parallel '/bin/time -v ${AKHAL_DIR}/akhal parse hg38.pggb.lcpan.l{}.rgfa >> hg38.pggb.lcpan.l{}.out 2>&1' ::: 4 5 6 7
    parallel '${AKHAL_DIR}/akhal stats hg38.pggb.lcpan.l{}.rgfa >> hg38.pggb.lcpan.l{}.out 2>&1' ::: 4 5 6 7

    for l in 4 5 6 7; do
        parse_2files "hg38.pggb.lcpan.l${l}.rgfa" "hg38.pggb.lcpan.l${l}.out" "stats.txt"
    done

    rm -f *.log
    rm -f *.rgfa

    cd ..
fi

### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------
### LCP levels analyses based on non-overlapping gfa graph
### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------

if [ "$LCPAN_GFA" = "true" ]; then
    echo "Experiment on different LCP levels started (nov-gfa)..."

    mkdir -p lcpan-levels-nov-gfa
    cd lcpan-levels-nov-gfa

    for l in 4 5 6 7; do \
        /bin/time -v ${LCPAN_DIR}/lcpan -vg \
            -r "$FASTA" \
            -v "$HPRC" \
            -p hg38.pggb.lcpan.l${l} \
            -l ${l} \
            --verbose \
            --gfa > hg38.pggb.lcpan.l${l}.out 2>&1; \
        /bin/time -v bash ${LCPANMERGE_DIR}/lcpan-merge.sh hg38.pggb.lcpan.l${l}.log >> hg38.pggb.lcpan.l${l}.out 2>&1; \
    done
    parallel '/bin/time -v ${AKHAL_DIR}/akhal parse hg38.pggb.lcpan.l{}.gfa >> hg38.pggb.lcpan.l{}.out 2>&1' ::: 4 5 6 7
    parallel '${AKHAL_DIR}/akhal stats hg38.pggb.lcpan.l{}.gfa >> hg38.pggb.lcpan.l{}.out 2>&1' ::: 4 5 6 7

    for l in 4 5 6 7; do
        parse_2files "hg38.pggb.lcpan.l${l}.gfa" "hg38.pggb.lcpan.l${l}.out" "stats.txt"
    done

    rm -f *.log
    rm -f *.gfa

    cd ..
fi

### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------
### Thread scaling analyses based on non-overlapping gfa graph
### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------

if [ "$LCPAN_THD" = "true" ]; then
    echo "Experiment on different thread numbers started (nov-gfa)..."

    mkdir -p lcpan-threads
    cd lcpan-threads

    for t in 1 2 4 8 16; do \
        /bin/time -v ${LCPAN_DIR}/lcpan -vg \
            -r "$FASTA" \
            -v "$HPRC" \
            -p hg38.pggb.lcpan.t${t} \
            -t $t \
            --gfa \
            --verbose > hg38.pggb.lcpan.t${t}.out 2>&1; \
        /bin/time -v bash ${LCPANMERGE_DIR}/lcpan-merge.sh hg38.pggb.lcpan.t${t}.log >> hg38.pggb.lcpan.t${t}.out 2>&1; \
    done
    export AKHAL_DIR
    parallel '/bin/time -v ${AKHAL_DIR}/akhal parse hg38.pggb.lcpan.t{}.gfa >> hg38.pggb.lcpan.t{}.out 2>&1' ::: 1 2 4 8 16

    for t in 1 2 4 8 16; do
        parse_2files "hg38.pggb.lcpan.t${t}.gfa" "hg38.pggb.lcpan.t${t}.out" "stats.txt"
    done

    rm -f *.log
    rm -f *.gfa

    cd ..
fi

### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------
### Thread scaling analyses based on VG tools
### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------

if [ "$VG_THD" = "true" ]; then
    echo "Experiment on different thread numbers started (vg)..."

    mkdir -p vg-threads
    cd vg-threads

    CHUNK_SIZE=10000000
    rm -f hg38.chunks.txt

    # Construct chunks
    while read -r chr chr_length _; do \
        for ((i = 0; i * CHUNK_SIZE < chr_length; i++)); do \
            start=$((i * CHUNK_SIZE + 1)); \
            end=$(((i + 1) * CHUNK_SIZE)); \
            [[ $end -gt $chr_length ]] && end=$chr_length; \
            echo "${chr}:${start}-${end}" >> hg38.chunks.txt; \
        done; \
    done < "$FAI"

    # Run threads
    for t in 1 2 4 8 16; do
        /bin/time -v bash -c '
            i=0
            while read -r region; do
                output_file="chunk.${i}.vg"
                vg construct -r "$1" \
                            -v "$2" \
                            -f -R "$region" \
                            -t "$3" > "$output_file"
                ((i++))
            done < hg38.chunks.txt
        ' _ "$FASTA" "$HPRC_GZ" "$t" > "hg38.pggb.vg.t$t.out" 2>&1

        /bin/time -v vg combine -p chunk.*.vg > hg38.pggb.vg 2>> "hg38.pggb.vg.t$t.out"

        rm -f chunk.*.vg

        if [ "$t" -eq 1 ]; then
            /bin/time -v vg convert -f hg38.pggb.vg > hg38.pggb.vg.gfa 2>> "hg38.pggb.vg.t$t.out"
            parse_2files "hg38.pggb.vg.gfa" "hg38.pggb.vg.t$t.out" "stats.txt"
            ${AKHAL_DIR}/akhal stats "hg38.pggb.vg.gfa" >> "hg38.pggb.vg.t$t.out" 2>&1
            rm -f hg38.pggb.vg.gfa
        else
            parse_2files "hg38.pggb.vg" "hg38.pggb.vg.t$t.out" "stats.txt"
        fi

        rm -f hg38.pggb.vg
    done

    rm -f hg38.chunks.txt

    cd ..
fi

### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------
### Alignment experiment (chr1_10_22)
### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------

if [ "$ALIGN_LIMITED" = "true" ]; then
    echo "Experiment on alignment for hg002 (chr1_10_22) ..."

    mkdir -p alignment-chr1-10-22
    cd alignment-chr1-10-22

    PREFIX="hg38.chr1_10_22"
    SUB_HPRC="hprc-v1.0-pggb.grch38.chr1_10_22.vcf"

    if [ ! -f "$PREFIX.fa" ]; then
        # Prepare reference and vcf files for chr 1, 10, and 22
        seqkit grep -n -p "chr1" -p "chr10" -p "chr22" $FASTA -o "$PREFIX.fa"
        samtools faidx "$PREFIX.fa"
    else
        echo "$PREFIX.fa already exists. skipping..."
    fi

    PREFIX_LEN=$(awk -F '\t' '{sum += $2} END {print sum}' "$PREFIX.fa.fai")
    fa_stats $PACBIO_SUB_FA $PREFIX_LEN "stats.txt"
    fa_stats $ONT_SUB_FA $PREFIX_LEN "stats.txt"

    if [ ! -f "$SUB_HPRC" ]; then
        grep -w '^#\|chr1\|chr10\|chr22' $HPRC > $SUB_HPRC
    else
        echo "$SUB_HPRC already exists. skipping..." 
    fi

    ### Create variation graphs
    if [ ! -f "$PREFIX.pggb.lcpan.gfa" ]; then
        echo "LCPan graph construction for alignment started ..."
        /bin/time -v ${LCPAN_DIR}/lcpan -vg -r "$PREFIX.fa" -v $SUB_HPRC -p "$PREFIX.pggb.lcpan" --gfa --verbose > "$PREFIX.pggb.lcpan.out" 2>&1
        /bin/time -v bash ${LCPANMERGE_DIR}/lcpan-merge.sh "$PREFIX.pggb.lcpan.log" >> "$PREFIX.pggb.lcpan.out" 2>&1
        ${AKHAL_DIR}/akhal stats "$PREFIX.pggb.lcpan.gfa" >> "$PREFIX.pggb.lcpan.out" 2>&1
        rm -f $PREFIX.pggb.lcpan.log
    else
        echo "$PREFIX.pggb.lcpan.gfa already exists. skipping construction..."
    fi 

    if [ ! -f "$PREFIX.pggb.vg.gfa" ]; then
        echo "VG graph construction for alignment started ..."
        /bin/time -v vg construct -r "$PREFIX.fa" -v $SUB_HPRC -m 64 -N -S > "$PREFIX.pggb.vg" 2> "$PREFIX.pggb.vg.out"
        /bin/time -v vg convert -f "$PREFIX.pggb.vg" > "$PREFIX.pggb.vg.gfa" 2>> "$PREFIX.pggb.vg.out"
        ${AKHAL_DIR}/akhal stats "$PREFIX.pggb.vg.gfa" >> "$PREFIX.pggb.vg.out" 2>&1
        rm -f "$PREFIX.pggb.vg"
    else
        echo "$PREFIX.pggb.vg.gfa already exists. skipping construction..."
    fi

    # Only needed for construction of vg graphs
    rm -f $SUB_HPRC

    ## Align HiFi reads of (hg002)
    echo "PacBio-HiFi GraphAligner"
    graphaligner_run "lcpan" "$PACBIO_SUB_FA" "hifi" "$PREFIX" "hg002.chr1_10_22"
    graphaligner_run "vg" "$PACBIO_SUB_FA" "hifi" "$PREFIX" "hg002.chr1_10_22"

    ## Align ONT reads of (hg002)
    echo "ONT GraphAligner"
    graphaligner_run "lcpan" "$ONT_SUB_FA" "ont" "$PREFIX" "hg002.chr1_10_22"
    graphaligner_run "vg" "$ONT_SUB_FA" "ont" "$PREFIX" "hg002.chr1_10_22"

    mkdir -p out
    mv *.out out

    if [ ! -f "HG002_GRCh38.chr1_10_22.pbsv.expanded.bed" ]; then
        bcftools view -r chr1,chr10,chr22 "$HG002_VCF" -o HG002_GRCh38.chr1_10_22.pbsv.vcf -O v
        bcftools query -f '%CHROM\t%POS0\t%END\n' HG002_GRCh38.chr1_10_22.pbsv.vcf > HG002_GRCh38.chr1_10_22.pbsv.bed
        rm -f HG002_GRCh38.chr1_10_22.pbsv.vcf

        awk '{print $1"\t"$2}' "$PREFIX.fa.fai" > "$PREFIX.txt"
        bedtools slop -i HG002_GRCh38.chr1_10_22.pbsv.bed -g "$PREFIX.txt" -b 100 > HG002_GRCh38.chr1_10_22.pbsv.expanded.bed
        rm -f "$PREFIX.txt" HG002_GRCh38.chr1_10_22.pbsv.bed
    fi

    echo -e "Method\tTP\tFP\tFN\tPrecision\tRecall\tF1" >> "stats.txt"

    ## Pacbio-HiFi SV detection
    echo "Pacbio-HiFi pbsv"
    pbsv_run "lcpan" "$PACBIO_SUB_FA" "hifi" "$PREFIX" "hg002.chr1_10_22"
    pbsv_run "vg" "$PACBIO_SUB_FA" "hifi" "$PREFIX" "hg002.chr1_10_22"

    ## ONT SV detection
    echo "ONT pbsv"
    pbsv_run "lcpan" "$ONT_SUB_FA" "ont" "$PREFIX" "hg002.chr1_10_22"
    pbsv_run "vg" "$ONT_SUB_FA" "ont" "$PREFIX" "hg002.chr1_10_22"

    rm -f *.bed

    mv *.out out

    cd ..
fi

### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------
### Alignment experiment
### ---------------------------------------------------------------------------------
### ---------------------------------------------------------------------------------

if [ "$ALIGN" = "true" ]; then
    echo "Experiment on alignment for hg002 ..."

    mkdir -p alignment
    cd alignment
    
    PREFIX="human_v38"

    TOTAL_LEN=$(awk -F '\t' '{sum += $2} END {print sum}' "$FAI")
    fa_stats $PACBIO_FA $TOTAL_LEN "stats.txt"

    ### Create variation graphs
    if [ ! -f "$PREFIX.pggb.lcpan.gfa" ]; then
        echo "LCPan graph construction for alignment started ..."
        /bin/time -v ${LCPAN_DIR}/lcpan -vg -r $FASTA -v $HPRC -p "$PREFIX.pggb.lcpan" --gfa --verbose > "$PREFIX.pggb.lcpan.out" 2>&1
        /bin/time -v bash ${LCPANMERGE_DIR}/lcpan-merge.sh "$PREFIX.pggb.lcpan.log" >> "$PREFIX.pggb.lcpan.out" 2>&1
        ${AKHAL_DIR}/akhal stats "$PREFIX.pggb.lcpan.gfa" >> "$PREFIX.pggb.lcpan.out" 2>&1
        rm -f $PREFIX.pggb.lcpan.log
    else
        echo "$PREFIX.pggb.lcpan.gfa already exists. skipping construction..."
    fi

    if [ ! -f "$PREFIX.pggb.vg.gfa" ]; then
        echo "VG graph construction for alignment started ..."
        /bin/time -v vg construct -r "$FASTA" -v $HPRC -m 64 -N -S > "$PREFIX.pggb.vg" 2> "$PREFIX.pggb.vg.out"
        /bin/time -v vg convert -f "$PREFIX.pggb.vg" > "$PREFIX.pggb.vg.gfa" 2>> "$PREFIX.pggb.vg.out"
        ${AKHAL_DIR}/akhal stats "$PREFIX.pggb.vg.gfa" >> "$PREFIX.pggb.vg.out" 2>&1
        rm -f "$PREFIX.pggb.vg"
    else
        echo "$PREFIX.pggb.lcpan.gfa already exists. skipping construction..."
    fi

    #### Align HiFi reads of (hg002)
    echo "PacBio-HiFi GraphAligner"
    graphaligner_run "lcpan" "$PACBIO_FA" "hifi" "$PREFIX" "hg002"
    graphaligner_run "vg" "$PACBIO_FA" "hifi" "$PREFIX" "hg002"

    mkdir -p out
    mv *.out out

    if [ ! -f "HG002_GRCh38.pbsv.expanded.bed" ]; then
        bcftools query -f '%CHROM\t%POS0\t%END\n' $HG002_VCF > HG002_GRCh38.pbsv.bed

        awk '{print $1"\t"$2}' "$PREFIX.fa.fai" > "$PREFIX.txt"
        bedtools slop -i HG002_GRCh38.pbsv.bed -g "$PREFIX.txt" -b 100 > HG002_GRCh38.pbsv.expanded.bed
        rm -f "$PREFIX.txt" HG002_GRCh38.pbsv.bed
    fi

    echo -e "Method\tTP\tFP\tFN\tPrecision\tRecall\tF1" >> "stats.txt"

    ## Pacbio-HiFi SV detection
    echo "Pacbio-HiFi pbsv"
    pbsv_run "lcpan" "$PACBIO_FA" "hifi" "$PREFIX" "hg002"
    pbsv_run "vg" "$PACBIO_FA" "hifi" "$PREFIX" "hg002"

    rm -f *.bed

    mv *.out out

    cd ..
fi
